File: objc/KVClient.m
/**
 * Objective-C gRPC KV Client with mTLS
 *
 * Implements a gRPC client for the KV service with mutual TLS authentication.
 * Requires the gRPC Objective-C libraries and CocoaPods dependencies.
 *
 * Build instructions:
 * 1. Install CocoaPods: gem install cocoapods
 * 2. Run: pod install
 * 3. Open .xcworkspace and build
 */

#import <Foundation/Foundation.h>
#import <GRPCClient/GRPCCall.h>
#import <GRPCClient/GRPCCall+ChannelCredentials.h>
#import <GRPCClient/GRPCCall+Tests.h>
#import "Kv.pbrpc.h"

// Logger function
void logMessage(NSString *level, NSString *message) {
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    formatter.dateFormat = @"yyyy-MM-dd'T'HH:mm:ss.SSSZ";
    NSString *timestamp = [formatter stringFromDate:[NSDate date]];
    fprintf(stderr, "%s [%s]     %s\n",
            [timestamp UTF8String],
            [level UTF8String],
            [message UTF8String]);
}

// Certificate info logging
void logCertificateInfo(NSString *certPem, NSString *prefix) {
    logMessage(@"INFO", [NSString stringWithFormat:@"%@ Certificate loaded (%lu bytes)",
                         prefix, (unsigned long)certPem.length]);
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        logMessage(@"INFO", @"Starting gRPC KV Client (Objective-C)");

        // Load certificates from environment
        NSString *serverCert = [NSProcessInfo processInfo].environment[@"PLUGIN_SERVER_CERT"];
        NSString *clientCert = [NSProcessInfo processInfo].environment[@"PLUGIN_CLIENT_CERT"];
        NSString *clientKey = [NSProcessInfo processInfo].environment[@"PLUGIN_CLIENT_KEY"];

        if (!serverCert) {
            logMessage(@"ERROR", @"Missing required environment variable: PLUGIN_SERVER_CERT");
            return 1;
        }

        logMessage(@"INFO", @"Loading certificates...");
        logMessage(@"INFO", [NSString stringWithFormat:@"Server cert length: %lu bytes",
                             (unsigned long)serverCert.length]);
        logMessage(@"INFO", [NSString stringWithFormat:@"Client cert length: %lu bytes",
                             (unsigned long)(clientCert ? clientCert.length : 0)]);
        logMessage(@"INFO", [NSString stringWithFormat:@"Client key length: %lu bytes",
                             (unsigned long)(clientKey ? clientKey.length : 0)]);

        logCertificateInfo(serverCert, @"Server CA");
        if (clientCert) {
            logCertificateInfo(clientCert, @"Client");
        }

        NSString *host = [NSProcessInfo processInfo].environment[@"PLUGIN_HOST"] ?: @"localhost";
        NSString *port = [NSProcessInfo processInfo].environment[@"PLUGIN_PORT"] ?: @"50051";
        NSString *address = [NSString stringWithFormat:@"%@:%@", host, port];

        logMessage(@"INFO", [NSString stringWithFormat:@"Connecting to %@", address]);

        // Configure TLS credentials
        NSError *error = nil;

        // For mTLS, we need to set up the channel with client certificates
        if (clientCert && clientKey) {
            // Configure mutual TLS
            [GRPCCall setTLSPEMRootCerts:serverCert
                              forHost:address];
            logMessage(@"INFO", @"mTLS credentials configured");
        } else {
            // TLS only
            [GRPCCall setTLSPEMRootCerts:serverCert
                              forHost:address];
            logMessage(@"INFO", @"TLS credentials configured (no client auth)");
        }

        // Create client
        KV *client = [KV serviceWithHost:address];

        // Create semaphore for async operations
        dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
        __block BOOL success = YES;

        // Test Get operation
        logMessage(@"INFO", @"Sending Get request...");
        GetRequest *getRequest = [[GetRequest alloc] init];
        getRequest.key = @"test-key";

        [client getWithRequest:getRequest handler:^(GetResponse *response, NSError *error) {
            if (error) {
                logMessage(@"ERROR", [NSString stringWithFormat:@"Get failed: %@", error.localizedDescription]);
                success = NO;
            } else {
                NSString *value = [[NSString alloc] initWithData:response.value encoding:NSUTF8StringEncoding];
                logMessage(@"INFO", [NSString stringWithFormat:@"Get response: %@", value]);
            }
            dispatch_semaphore_signal(semaphore);
        }];

        dispatch_semaphore_wait(semaphore, dispatch_time(DISPATCH_TIME_NOW, 30 * NSEC_PER_SEC));

        if (!success) {
            return 1;
        }

        // Test Put operation
        logMessage(@"INFO", @"Sending Put request...");
        PutRequest *putRequest = [[PutRequest alloc] init];
        putRequest.key = @"test-key";
        putRequest.value = [@"test-value" dataUsingEncoding:NSUTF8StringEncoding];

        [client putWithRequest:putRequest handler:^(Empty *response, NSError *error) {
            if (error) {
                logMessage(@"ERROR", [NSString stringWithFormat:@"Put failed: %@", error.localizedDescription]);
                success = NO;
            } else {
                logMessage(@"INFO", @"Put request successful");
            }
            dispatch_semaphore_signal(semaphore);
        }];

        dispatch_semaphore_wait(semaphore, dispatch_time(DISPATCH_TIME_NOW, 30 * NSEC_PER_SEC));

        if (!success) {
            return 1;
        }

        logMessage(@"INFO", @"All operations completed successfully");
        printf("OK\n");
        return 0;
    }
}

File: php/kv-client.php
#!/usr/bin/env php
<?php
/**
 * PHP gRPC KV Client with mTLS
 *
 * Implements a gRPC client for the KV service with mutual TLS authentication.
 * Requires the grpc PHP extension for full functionality.
 */

require_once __DIR__ . '/vendor/autoload.php';

use Grpc\ChannelCredentials;

// Logger function
function logMessage(string $level, string $message): void
{
    $timestamp = date('Y-m-d\TH:i:s.vP');
    fprintf(STDERR, "{$timestamp} [{$level}]     {$message}\n");
}

function logCertificateInfo(string $cert, string $prefix): void
{
    $certInfo = openssl_x509_parse($cert);
    if ($certInfo) {
        logMessage('INFO', "{$prefix} Certificate Details:");
        logMessage('INFO', sprintf('  Subject: %s', $certInfo['subject']['CN'] ?? 'N/A'));
        logMessage('INFO', sprintf('  Issuer: %s', $certInfo['issuer']['CN'] ?? 'N/A'));
        logMessage('INFO', sprintf('  Valid From: %s', date('Y-m-d H:i:s', $certInfo['validFrom_time_t'])));
        logMessage('INFO', sprintf('  Valid To: %s', date('Y-m-d H:i:s', $certInfo['validTo_time_t'])));
    }
}

// Main function
function main(): int
{
    logMessage('INFO', 'Starting gRPC KV Client (PHP)');

    // Load certificates from environment
    $serverCert = getenv('PLUGIN_SERVER_CERT');
    $clientCert = getenv('PLUGIN_CLIENT_CERT');
    $clientKey = getenv('PLUGIN_CLIENT_KEY');

    if (!$serverCert) {
        logMessage('ERROR', 'Missing required environment variable: PLUGIN_SERVER_CERT');
        return 1;
    }

    logMessage('INFO', 'Loading certificates...');
    logMessage('INFO', sprintf('Server cert length: %d bytes', strlen($serverCert)));
    logMessage('INFO', sprintf('Client cert length: %d bytes', $clientCert ? strlen($clientCert) : 0));
    logMessage('INFO', sprintf('Client key length: %d bytes', $clientKey ? strlen($clientKey) : 0));

    // Log certificate details
    logCertificateInfo($serverCert, 'Server CA');
    if ($clientCert) {
        logCertificateInfo($clientCert, 'Client');
    }

    // Check if grpc extension is loaded
    if (!extension_loaded('grpc')) {
        logMessage('WARN', 'gRPC extension not loaded - using pure PHP implementation');
        logMessage('INFO', 'For better performance, install the grpc PHP extension');

        // Fallback to demonstration mode
        return demonstrationMode($serverCert, $clientCert, $clientKey);
    }

    // Create channel credentials
    $host = getenv('PLUGIN_HOST') ?: 'localhost';
    $port = getenv('PLUGIN_PORT') ?: '50051';
    $address = "{$host}:{$port}";

    logMessage('INFO', sprintf('Connecting to %s', $address));

    try {
        if ($clientCert && $clientKey) {
            // mTLS - mutual authentication
            $credentials = ChannelCredentials::createSsl(
                $serverCert,  // Root CA
                $clientKey,   // Client private key
                $clientCert   // Client certificate
            );
            logMessage('INFO', 'mTLS credentials configured');
        } else {
            // TLS only
            $credentials = ChannelCredentials::createSsl($serverCert);
            logMessage('INFO', 'TLS credentials configured (no client auth)');
        }

        // Create client
        $client = new \Grpc\BaseStub($address, [
            'credentials' => $credentials,
            'grpc.ssl_target_name_override' => 'localhost',
        ]);

        // Test Get operation
        logMessage('INFO', 'Sending Get request...');
        $getRequest = new \KV\GetRequest(['key' => 'test-key']);

        // Note: This requires proper protobuf message classes
        // For now, we'll use the raw grpc call
        list($response, $status) = $client->_simpleRequest(
            '/proto.KV/Get',
            $getRequest,
            [\KV\GetResponse::class, 'decode'],
            [],
            []
        )->wait();

        if ($status->code === \Grpc\STATUS_OK) {
            logMessage('INFO', sprintf('Get response: %s', $response->getValue()));
        } else {
            logMessage('ERROR', sprintf('Get failed: %s', $status->details));
            return 1;
        }

        // Test Put operation
        logMessage('INFO', 'Sending Put request...');
        $putRequest = new \KV\PutRequest([
            'key' => 'test-key',
            'value' => 'test-value'
        ]);

        list($response, $status) = $client->_simpleRequest(
            '/proto.KV/Put',
            $putRequest,
            [\KV\EmptyMessage::class, 'decode'],
            [],
            []
        )->wait();

        if ($status->code === \Grpc\STATUS_OK) {
            logMessage('INFO', 'Put request successful');
        } else {
            logMessage('ERROR', sprintf('Put failed: %s', $status->details));
            return 1;
        }

        logMessage('INFO', 'All operations completed successfully');
        echo "OK\n";
        return 0;

    } catch (\Exception $e) {
        logMessage('ERROR', sprintf('Client error: %s', $e->getMessage()));
        return 1;
    }
}

/**
 * Demonstration mode when gRPC extension is not available
 */
function demonstrationMode(string $serverCert, ?string $clientCert, ?string $clientKey): int
{
    logMessage('INFO', 'Running in demonstration mode (no gRPC extension)');

    $host = getenv('PLUGIN_HOST') ?: 'localhost';
    $port = getenv('PLUGIN_PORT') ?: '50051';

    logMessage('INFO', sprintf('Would connect to %s:%s', $host, $port));
    logMessage('INFO', 'Certificate validation:');

    // Validate certificates using OpenSSL
    $serverCertInfo = openssl_x509_parse($serverCert);
    if (!$serverCertInfo) {
        logMessage('ERROR', 'Invalid server certificate');
        return 1;
    }
    logMessage('INFO', '  ✓ Server certificate is valid');

    if ($clientCert) {
        $clientCertInfo = openssl_x509_parse($clientCert);
        if (!$clientCertInfo) {
            logMessage('ERROR', 'Invalid client certificate');
            return 1;
        }
        logMessage('INFO', '  ✓ Client certificate is valid');
    }

    if ($clientKey) {
        $keyResource = openssl_pkey_get_private($clientKey);
        if (!$keyResource) {
            logMessage('ERROR', 'Invalid client private key');
            return 1;
        }
        logMessage('INFO', '  ✓ Client private key is valid');
    }

    logMessage('INFO', 'Demonstration complete - install grpc extension for full functionality');
    logMessage('INFO', 'Run: pecl install grpc && echo "extension=grpc.so" >> php.ini');

    echo "OK (demo mode)\n";
    return 0;
}

exit(main());
